File: javascript/security_issues.js
/**
 * This file contains intentional security issues for testing security analysis tools.
 * DO NOT use this code in production!
 */

// Hardcoded credentials (Security issue: Hardcoded secret)
const DB_PASSWORD = 'mySuperSecretPassword123!';
const API_KEYS = {
  production: 'prod_1234567890abcdef',
  staging: 'staging_0987654321',
  test: 'test_key_abcdef123456',
  // More hardcoded credentials
  aws: {
    accessKeyId: 'AKIAIOSFODNN7EXAMPLE',
    secretAccessKey: 'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY',
    region: 'us-east-1'
  },
  // JWT secret (should be in environment variables)
  jwtSecret: 'my_jwt_super_secret_key_12345!@#',
  // Database connection string with credentials
  dbConnection: 'mongodb://admin:admin123@localhost:27017/production'
};

// SQL Injection vulnerability
function getUserDetails(username) {
  // WARNING: This is vulnerable to SQL injection!
  const query = `SELECT * FROM users WHERE username = '${username}'`;
  return executeQuery(query);
}

// Command Injection vulnerability
function pingHost(host) {
  // WARNING: This is vulnerable to command injection!
  const { exec } = require('child_process');
  return new Promise((resolve, reject) => {
    exec(`ping -c 4 ${host}`, (error, stdout, stderr) => {
      if (error) {
        return reject(error);
      }
      resolve(stdout);
    });
  });
}

// XSS Vulnerability
function displayUserComment(comment) {
  // WARNING: This is vulnerable to XSS!
  document.getElementById('comment-section').innerHTML = comment;
}

// Insecure random number generation
function generateSessionToken() {
  // WARNING: Math.random() is not cryptographically secure!
  return Math.random().toString(36).substring(2);
}

// Sensitive data in logs
function logUserActivity(userId, action) {
  // WARNING: Logging sensitive information!
  console.log(`User ${userId} performed action: ${action} with token: ${API_KEYS.production}`);
}

// Insecure file operations
const fs = require('fs');
const path = require('path');

function readUserFile(username, filename) {
  // WARNING: Path traversal vulnerability!
  const filePath = path.join('/data/users', username, filename);
  return fs.readFileSync(filePath, 'utf8');
}

// Insecure deserialization
function deserializeUserData(serializedData) {
  // WARNING: eval() can execute arbitrary code!
  try {
    return eval('(' + serializedData + ')');
  } catch (e) {
    console.error('Deserialization failed:', e);
    return null;
  }
}

// Insecure direct object reference (IDOR) vulnerability
function getUserProfile(userId) {
  // WARNING: No authorization check!
  return db.users.findOne({ id: userId });
}

// Insecure cookie settings (missing httpOnly and secure flags)
function setSessionCookie(res, userId) {
  res.cookie('session', userId, {
    maxAge: 24 * 60 * 60 * 1000, // 1 day
    httpOnly: false, // Should be true
    secure: false,   // Should be true in production
    sameSite: 'lax'
  });
}

// SQL Injection vulnerability (using string concatenation)
function searchUsers(searchTerm) {
  // WARNING: Vulnerable to SQL injection!
  return db.query(`SELECT * FROM users WHERE username LIKE '%${searchTerm}%'`);
}

// Hardcoded encryption key (Security issue)
const ENCRYPTION_KEY = 'insecure-static-key-12345';

// Insecure password hashing (using weak algorithm)
function hashPassword(password) {
  const crypto = require('crypto');
  // WARNING: Using weak MD5 hashing
  return crypto.createHash('md5').update(password).digest('hex');
}

// Insecure random number generation
function generateResetToken() {
  // WARNING: Math.random() is not cryptographically secure!
  return Math.random().toString(36).substring(2, 15) + 
         Math.random().toString(36).substring(2, 15);
}

// Insecure cookie settings
function setAuthCookie(res, userId) {
  // WARNING: Insecure cookie settings!
  res.cookie('auth', userId, {
    httpOnly: false, // Should be true
    secure: false,   // Should be true in production
    sameSite: 'none',
    maxAge: 30 * 24 * 60 * 60 * 1000 // 30 days
  });
}

// Hardcoded encryption key (Security issue: Insecure key management)
const CRYPTO_KEY = 'insecureStaticKey123!';

// Insecure encryption
function encryptData(data) {
  const crypto = require('crypto');
  const cipher = crypto.createCipher('aes-256-cbc', CRYPTO_KEY);
  let encrypted = cipher.update(JSON.stringify(data), 'utf8', 'hex');
  encrypted += cipher.final('hex');
  return encrypted;
}

// Example of prototype pollution vulnerability
function merge(target, source) {
  // WARNING: This is vulnerable to prototype pollution!
  for (const key in source) {
    target[key] = source[key];
  }
  return target;
}

// Example usage of vulnerable functions
const userInput = req.query.username || 'guest';
const user = getUserDetails(userInput);

// Log sensitive information
logUserActivity(user.id, 'login');

// Set insecure cookie
setAuthCookie(res, user.id);

// Example of IDOR vulnerability
app.get('/api/user/profile/:id', (req, res) => {
  // No authorization check - any user can access any profile!
  const userProfile = getUserProfile(req.params.id);
  res.json(userProfile);
});

// Example of SQL injection vulnerability
app.get('/api/users/search', (req, res) => {
  const results = searchUsers(req.query.term);
  res.json(results);
});

// Example of insecure password reset
app.post('/api/reset-password', (req, res) => {
  const token = generateResetToken();
  // In a real app, this would be sent via email
  console.log(`Password reset token for ${req.body.email}: ${token}`);
  res.json({ success: true });
});

// Example of insecure file upload
const multer = require('multer');
const upload = multer({ dest: 'uploads/' });

app.post('/api/upload', upload.single('file'), (req, res) => {
  // WARNING: No file type validation!
  res.json({ success: true, file: req.file });
});

// Export for testing
module.exports = {
  getUserDetails,
  pingHost,
  displayUserComment,
  generateSessionToken,
  readUserFile,
  deserializeUserData,
  encryptData,
  merge
};

File: python/security_issues.py
"""
This file contains intentional security issues for testing security analysis tools.
"""
import os
import subprocess
import pickle
import base64

# Hardcoded credentials (Security issue: Hardcoded secret)
API_KEY = "s3cr3t_k3y_12345"
DB_PASSWORD = "db@admin#pass"

# Command injection vulnerability
def run_command(user_input):
    """Run a command with user input (vulnerable to command injection)."""
    return subprocess.check_output(f"echo {user_input}", shell=True)  # Shell injection vulnerability

# SQL injection vulnerability
def query_database(query):
    """Execute a database query (vulnerable to SQL injection)."""
    import sqlite3
    conn = sqlite3.connect('example.db')
    cursor = conn.cursor()
    cursor.execute(f"SELECT * FROM users WHERE username = '{query}'")
    return cursor.fetchall()

# Insecure deserialization
def deserialize_data(serialized_data):
    """Deserialize data (vulnerable to arbitrary code execution)."""
    return pickle.loads(base64.b64decode(serialized_data))

# Hardcoded encryption key (Security issue: Insecure key management)
ENCRYPTION_KEY = b'ThisIsAVeryInsecureKey123!'

def encrypt_data(data):
    """Encrypt data with hardcoded key (insecure implementation)."""
    from cryptography.fernet import Fernet
    cipher_suite = Fernet(ENCRYPTION_KEY)
    return cipher_suite.encrypt(data.encode())

# Sensitive data in logs
def log_sensitive_info():
    """Log sensitive information (Security issue: Information exposure)."""
    import logging
    logging.basicConfig(level=logging.INFO)
    logging.info(f"User logged in with token: {API_KEY}")

# Insecure temporary file
def create_temp_file():
    """Create a temporary file (vulnerable to race conditions)."""
    import tempfile
    temp = tempfile.mktemp()
    with open(temp, 'w') as f:
        f.write("Temporary data")
    return temp

def hash_password_insecure(password):
    """
    Insecure password hashing using MD5 (vulnerable to rainbow table attacks).
    WARNING: This is intentionally vulnerable for testing purposes.
    """
    import hashlib
    # Using MD5 which is cryptographically broken (Security issue)
    return hashlib.md5(password.encode()).hexdigest()

def verify_jwt_token(token):
    """
    Insecure JWT verification (vulnerable to signature bypass).
    WARNING: This is intentionally vulnerable for testing purposes.
    """
    import jwt
    
    # Hardcoded secret key (Security issue)
    JWT_SECRET = 'insecure_secret_key_12345'
    
    try:
        # Decode without verifying the signature (Security issue)
        decoded = jwt.decode(token, options={"verify_signature": False})
        
        # Verify expiration manually (but signature is already not verified)
        if 'exp' in decoded and decoded['exp'] < int(time.time()):
            print("Token has expired")
            return None
            
        return decoded
    except Exception as e:
        print(f"JWT verification error: {str(e)}")
        return None

if __name__ == "__main__":
    # Demonstrate command injection
    user_input = input("Enter your name: ")
    print(run_command(user_input))
    
    # Log sensitive information
    log_sensitive_info()
    
    # Create insecure temporary file
    temp_file = create_temp_file()
    print(f"Created temporary file: {temp_file}")
    
    # Test JWT verification (vulnerable)
    test_token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImFkbWluIiwiaWF0IjoxNTE2MjM5MDIyfQ.XbPfbIHMI6arZ3Y922BhjWgQzWXcXNrz0ogtVhfEd2o"
    user_data = verify_jwt_token(test_token)
    print(f"Decoded user data: {user_data}")
    
    # Test insecure password hashing
    password = "myp@ssw0rd123"
    hashed = hash_password_insecure(password)
    print(f"Insecurely hashed password: {hashed}")
